Add max_single_primary_size as a condition for the rollover index API

joegallo opened the pull request:
Related to #63026

This PR ignores ILM and any deprecation that we might end up doing. It's just adding a new feature, the ability to specify `max_single_primary_size` as a condition for the `_rollover` API.

Note: I still have some docs and test changes to get in, but this is probably 90% done and would be okay to review.

joegallo committed:
Drop getAverageSizeInBytes

It had no non-test callers, so YAGNI.

joegallo committed:
Refactor TransportRolloverAction a bit

to separate the building of a Conditition.Stats cleanly from the
evaluation of same.

joegallo committed:
Whitespace

joegallo committed:
Client changes to keep up

joegallo committed:
Cleanups

Reorder things to be age, docs, size more consistently, and a few
'this way, not that way' changes.

joegallo committed:
Add MaxSinglePrimarySizeCondition

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indices/rollover/RolloverRequest.java:
```diff
@@ -116,6 +118,19 @@ public RolloverRequest addMaxIndexSizeCondition(ByteSizeValue size) {
+    /**
+     * Adds a size-based condition to check if the size of the largest primary shard is at least <code>size</code>.
+     */
+    public RolloverRequest addMaxIndexSinglePrimarySizeCondition(ByteSizeValue size) {
```
I think this can remove the "Index" part, so it can be `addMaxSinglePrimarySizeCondition`

joegallo replied:
Ehhhhh, I'm with you on disliking the name, but the pattern for the other conditions ends up with an `Index` in the middle -- is it worth breaking the apparent convention for these?

dakrone replied:
I don't think it's a convention as much as the others make "English" sense without the `Index` in them (it could be argued that "Size" would be fine without it, but that one will hopefully eventually be deprecated). I'm not super concerned, so if you'd rather keep the `Index` that's fine with me.

joegallo replied:
Renamed in d00e2287e4b9de6ae5fb1b900847439b74ca0242

dakrone commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -202,6 +214,16 @@ public void addMaxIndexSizeCondition(ByteSizeValue size) {
+    /**
+     * Adds a size-based condition to check if the size of the largest primary shard is at least <code>size</code>.
+     */
+    public void addMaxIndexSinglePrimarySizeCondition(ByteSizeValue size) {
```
Same here about how I think we can drop `Index` and just go with `addMaxSinglePrimarySizeCondition`

dakrone commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequestBuilder.java:
```diff
@@ -58,6 +58,11 @@ public RolloverRequestBuilder addMaxIndexSizeCondition(ByteSizeValue size){
         return this;
     }
 
+    public RolloverRequestBuilder addMaxIndexSinglePrimarySizeCondition(ByteSizeValue size){
```
Super minor nit
```suggestion
    public RolloverRequestBuilder addMaxIndexSinglePrimarySizeCondition(ByteSizeValue size) {
```

(and same comment about dropping `Index`)

joegallo replied:
Fixed the whitespace in af3f3e9 -- still holding off on the method rename for now.

dakrone commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -167,30 +171,47 @@ public void onFailure(Exception e) {
+                .filter(shard -> shard.getShardRouting().primary())
+                .map(ShardStats::getStats)
+                .mapToLong(shard -> shard.docs.getTotalSizeInBytes())
+                .max().orElse(-1);
```
We use -1 for the primary size, but 0 for a missing total size below, maybe we should use either -1 or 0 for both since they're both sizes?

dakrone replied:
I thought about this more, this should be 0 instead of -1, because we want `max_single_primary_size: 0` to be the same as an unconditional rolloverrr

joegallo replied:
f6fe3f8

dakrone reviewed (commented):
I left some minor comments, but this looks great!

joegallo committed:
Whitespace

joegallo committed:
Default to 0 rather than -1

joegallo commented:
@elasticmachine run elasticsearch-ci/bwc

joegallo committed:
Merge branch 'master' into into rollover-max-single-shard-size

joegallo committed:
s/MaxIndexSinglePrimarySize/MaxSinglePrimarySize/g

joegallo committed:
Docs, as well as docs and client tests

joegallo committed:
Add yml test

Just a copy/paste + search&replace on 30_max_size_condition.yml,
nothing special.

joegallo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into rollover-max-single-shard-size

joegallo committed:
Add "version:" restriction to the yml test for bwc

joegallo committed:
Add testRolloverMaxSinglePrimarySize

It's just a copy/paste/edit of testRolloverMaxSize, but it's still got
some value.

mark-vieira commented:
@elasticmachine update branch

joegallo committed:
Merge branch 'master' into rollover-max-single-shard-size

dakrone reviewed (approved):
LGTM, thanks for iterating on this!

joegallo commented:
Thanks for iterating with me!
